### ISSUE TITLE
Option to warn or error on unreachable statements/expressions

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2080]
**Is your feature request related to a problem? Please describe.**
In relation to #2079, right now the behaviour for Never cases isn't the nicest from an error standpoint, for me it lead to errors about the relevant variable's attributes being "unknown" (since it'd been resolved to Never) and this was a misdirection away from the actual issue which was that an instance check was being treated as unreachable.

**Describe the solution you'd like**
Similarly to MYPY, having an option to make Pyright warn or error whenever a statement or expression is resolved to Never (is unreachable) could help make the actual reason for errors like attributes resolving to unknown types more obvious and easy to understand.
 
**Additional context**
A simple example of attributes being "unknown" 
```
C:\Users\Lucina\PycharmProjects\tanjun\tanjun\parsing.py
  C:\Users\Lucina\PycharmProjects\tanjun\tanjun\parsing.py:630:17 - error: Type of "bind_client" is unknown (reportUnknownMemberType)
  C:\Users\Lucina\PycharmProjects\tanjun\tanjun\parsing.py:633:17 - error: Type of "bind_component" is unknown (reportUnknownMemberType)
```

for the following code

```py
if isinstance(converter, conversion.BaseConverter):
    if self._client:
        converter.bind_client(self._client)

    if self._component:
        converter.bind_component(self._component)
```

**Comment by erictraut** [comment]
There are many legitimate uses for unreachable code blocks in Python. I'm aware of the mypy check, but in my experience (and based on the number of mypy bugs filed on this topic), it gives so many false positive errors that it does not seem to provide a good tradeoff for most code.

Pyright already has a check called `reportUnnecessaryIsInstance` that will report an `isinstance` or `issubclass` check that is provably always true. It previously also checked for always-false conditions, but I needed to water down that check because of [this recent change](https://github.com/microsoft/pyright/issues/1851). That's why [the other bug you recently reported](https://github.com/microsoft/pyright/issues/2079) didn't generate any error, even in strict mode.

I'll need to think about this request a bit more. Could you provide a complete example (or two)? There's not enough context for me to understand the above code snippet.

**Comment by erictraut** [comment]
For now, I'm going to close this feature request. We may revisit it in the future if there's sufficient interest from other pyright users. Based on my experience, I think that the number of false positive errors would be way greater than the number of bugs it finds.

**Comment by ckuethe** [comment]
A trivial example:

```python

def do_something(arg:list[str]) - >str:
    if not isinstance(arg, list):
        raise ValueError("you did something silly.")
    return  ' '.join(arg)
```

I check early that a variable is of the expected type and throw my own error. The type hint is weak; maybe it helps me when I'm writing code (unless I just use notepad or nano) but it doesn't prevent a caller from passing in bad data before trying to process it. In this case I'm enforcing at runtime that the type is what I politely requested in the function signature.

... says the guy who just added a bunch of `# pyright: ignore reportUnnecessaryIsInstance` to various checks.

**Comment by erictraut** [comment]
Pyright version 1.1.402 adds support for a new diagnostic check called `reportUnreachable`. It reports cases where pyright detects code that is structurally unreachable (e.g. due to `return` or `raise` statements) or unreachable due to type analysis. This is an opt-in check and is not enabled by default even in strict mode.
